src/DotNetty.Common/ThreadLocalPool.cs

```diff
@@ -12,55 +12,375 @@ public class ThreadLocalPool
+                // reset our scavenge cursor
+                this.prevQueue = null;
+                this.cursorQueue = this.HeadQueue;
+                return false;
+            }
+
+            bool ScavengeSome()
+            {
+                WeakOrderQueue cursor = this.cursorQueue;
+                if (cursor == null)
+                {
+                    cursor = this.HeadQueue;
+                    if (cursor == null)
+                    {
+                        return false;
+                    }
+                }
+
+                bool success = false;
+                WeakOrderQueue prev = this.prevQueue;
+                do
+                {
+                    if (cursor.Transfer(this))
+                    {
+                        success = true;
+                        break;
+                    }
+
+                    WeakOrderQueue next = cursor.next;
+                    Thread ownerThread;
+                    if (!cursor.ownerThread.TryGetTarget(out ownerThread))
+                    {
+                        // If the thread associated with the queue is gone, unlink it, after
+                        // performing a volatile read to confirm there is no data left to collect.
+                        // We never unlink the first queue, as we don't want to synchronize on updating the head.
+                        if (!cursor.IsEmpty)
+                        {
+                            for (;;)
+                            {
+                                if (cursor.Transfer(this))
+                                {
+                                    success = true;
+                                }
+                                else
+                                {
+                                    break;
+                                }
+                            }
+                        }
+                        if (prev != null)
+                        {
+                            prev.next = next;
+                        }
+                    }
+                    else
+                    {
+                        prev = cursor;
+                    }
+
+                    cursor = next;
+                }
+                while (cursor != null && !success);
+
+                this.prevQueue = prev;
+                this.cursorQueue = cursor;
+                return success;
             }
         }
 
         internal static readonly int DefaultMaxCapacity = 262144;
         internal static readonly int InitialCapacity = Math.Min(256, DefaultMaxCapacity);
 
+        internal static readonly ThreadLocal<Dictionary<Stack, WeakOrderQueue>> DelayedPool
+            = new ThreadLocal<Dictionary<Stack, WeakOrderQueue>>(() => new Dictionary<Stack, WeakOrderQueue>());
+
         public ThreadLocalPool(int maxCapacity)
         {
             Contract.Requires(maxCapacity > 0);
-            this.MaxCapacity = maxCapacity;
+            this.MaxCapacity = Math.Max(0, maxCapacity);
```

Comment:
this contradicts to an Contract check above. Per the check maxCapacity can never be less than 0.